### PR TITLE
fix: Checking existence before calculating element uid

### PR DESCRIPTION
**Pull request by Dan-Maor**
It appears that the new implementation used for retrieving element UIDs causes issues in some areas, including when a find command is used while an application is during a transition stage.

The side effects include possible nil snapshots (and thus nil UIDs), blank snapshots (causing calculation of elements with bad attributes) and in some scenarios - unneeded slowdowns while XCTest tried to resolve elements which are no longer on the screen.

Example scenario:
* In the Settings app, find and click on a button element.
* Immediately search for all elements of type XCUIElementTypeButton.
This will cause snapshot operations to fail, and the resolve process will slow down and possibly return nil values.

Part of calling the full wdUID function checks the existence of the element prior to retrieving the UID, which is also needed in this case.

XCTest's implementation for `[XCUIElement exists]` calls a function called `[XCUIElementQuery allMatchingSnapshotsWithError:]` which returns an NSOrderedSet of snapshots, and if at least one snapshot exists - the element is considered existing. However this is not an indication that a unique snapshot exists for said element, and therefor when calling `resolveOrRaiseTestFailure` on an element a different function is used, called `[XCUIElementQuery uniqueMatchingSnapshotWithError:]`, which guarantees that the returned snapshot is unique for the element. It is slightly slower than `[XCUIElementQuery allMatchingSnapshotsWithError:]` due to additional verifications. In addition, `resolveOrRaiseTestFailure` also captures screenshots of the element and validate faults for later use - which is not needed in this scenario.

In order to avoid calling two snapshot retrievals for each element in the UID retrieval scenario (verifying `[XCUIElement exists]` and resolving it afterwards), the code now re-uses the snapshot retrieved from the function used in `resolveOrRaiseTestFailure` for both verifying existence (`[XCUIElementQuery uniqueMatchingSnapshotWithError:]` will return nil and assign an NSError) and for retrieving the UID information.

### CODE REVIEW

File: WebDriverAgentLib/Categories/XCUIElement+FBUID.m [hunk 1]
```diff
@@ -20,12 +21,19 @@ - (NSString *)fb_uid
...
+  // With Xcode 10, uniqueMatchingSnapshotWithError is not supported
   if (isSDKVersionLessThan(@"13.0")) {
     return self.fb_lastSnapshot.fb_uid;
   }
-  [self fb_nativeResolve];
-  return self.lastSnapshot.fb_uid;
+  NSError *error = nil;
+  // Using the Xcode 11 snapshot function used for resolving an element to validate existance and retrieve UID with the same snapshot
+  // Removes the need to take two snapshots (one for existance and one for resolving)
+  XCElementSnapshot *snapshot = [[self query] uniqueMatchingSnapshotWithError:&error];
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUID.m:35]: I'd rather check if query responds to `uniqueMatchingSnapshotWithError` and avoid checking the SDK version. also, we use our own `fb_query` wrapper.

**Comment by mykola-mokhnach** [comment]
Thanks for the investigation @Dan-Maor This is very useful info

File: PrivateHeaders/XCTest/XCUIElementQuery.h [hunk 2]
```diff
@@ -64,7 +64,7 @@
 - (XCElementSnapshot *)elementSnapshotForDebugDescription;
 // Added since Xcode 11.0
 - (XCElementSnapshot *)elementSnapshotForDebugDescriptionWithNoMatchesMessage:(id *)arg1;
-
+- (XCElementSnapshot*)uniqueMatchingSnapshotWithError:(id *)arg1;
```
Comment (mykola-mokhnach) [inline PrivateHeaders/XCTest/XCUIElementQuery.h]: Please add a comment about since when this method is available

Comment (mykola-mokhnach) [inline PrivateHeaders/XCTest/XCUIElementQuery.h]: perhaps `:(NSError **)error`

File: WebDriverAgentLib/Categories/XCUIElement+FBUID.m (hunk 1 above)
Comment (Dan-Maor) [reply WebDriverAgentLib/Categories/XCUIElement+FBUID.m:35]: I wasn't sure about using `includingNonModalElements` here, that's the main reason I didn't use `fb_query`. Also - I went in line with the `fb_lastSnapshot` which also does not use the wrapper.

File: WebDriverAgentLib/Categories/XCUIElement+FBUID.m [hunk 3]
```diff
@@ -20,12 +21,18 @@ - (NSString *)fb_uid
   if ([self respondsToSelector:@selector(accessibilityElement)]) {
     return [FBElementUtils uidWithAccessibilityElement:[self performSelector:@selector(accessibilityElement)]];
   }
-  // With Xcode 10, using fb_lastSnapshot is faster than resolving and using the lastSnapshot property
-  if (isSDKVersionLessThan(@"13.0")) {
-    return self.fb_lastSnapshot.fb_uid;
+  if ([self.query respondsToSelector:@selector(uniqueMatchingSnapshotWithError:)]) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUID.m]: I'd rather flip the if condition and put the immediate return there. the method would be easier to read in such case

Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBUID.m]: We could also cache the result into a static variable (taking into account this method is called quite often we might same some time for `respondsToSelector` verification)

File: WebDriverAgentLib/Categories/XCUIElement+FBUID.m (hunk 1 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBUID.m:35]: ok, makes sense

File: PrivateHeaders/XCTest/XCUIElementQuery.h [hunk 4]
```diff
@@ -64,7 +64,8 @@
 - (XCElementSnapshot *)elementSnapshotForDebugDescription;
 // Added since Xcode 11.0
 - (XCElementSnapshot *)elementSnapshotForDebugDescriptionWithNoMatchesMessage:(id *)arg1;
-
+// Added since Xcode 11.0
+- (XCElementSnapshot*)uniqueMatchingSnapshotWithError:(NSError **)arg1;
```
Comment (KazuCocoa) [inline PrivateHeaders/XCTest/XCUIElementQuery.h:68]: 👀

**Review by KazuCocoa — COMMENTED** [review]
thanks for the investigation

**Review by mykola-mokhnach — APPROVED** [review]

**Pull request merged**